src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,213 @@
+        _projectManager = projectManager;
+
+        _disposeTokenSource = new();
+        _workQueue = new(
+            enqueueDelay,
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    public void StartSending()
+    {
+        _active = true;
+
+        _projectManager.Changed += ProjectManager_Changed;
+
+        var projects = _projectManager.GetProjects();
+        foreach (var project in projects)
+        {
+            if (ProjectWorkspacePublishable(project))
+            {
+                // Don't enqueue project addition as those are unlikely to come through in large batches.
+                // Also ensures that we won't get project removal go through the queue without project addition.
+                ImmediatePublish(project, _disposeTokenSource.Token);
+            }
+        }
+    }
+
+    private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+    {
+        // Don't do any work if the solution is closing
+        if (args.SolutionIsClosing)
+        {
+            return;
+        }
+
+        switch (args.Kind)
+        {
+            case ProjectChangeKind.ProjectChanged:
+            case ProjectChangeKind.DocumentRemoved:
+            case ProjectChangeKind.DocumentAdded:
+                if (!ProjectWorkspacePublishable(args.Newer))
+                {
+                    break;
+                }
+
+                EnqueuePublish(args.Newer.AssumeNotNull());
+
+                break;
+
+            case ProjectChangeKind.ProjectAdded:
+
+                if (ProjectWorkspacePublishable(args.Newer))
+                {
+                    // Don't enqueue project addition as those are unlikely to come through in large batches.
+                    // Also ensures that we won't get project removal go through the queue without project addition.
+                    ImmediatePublish(args.Newer.AssumeNotNull(), _disposeTokenSource.Token);
+                }
+
+                break;
+
+            case ProjectChangeKind.ProjectRemoved:
+                // Enqueue removal so it will replace any other project changes in the work queue as they unnecessary now.
+                EnqueueRemoval(args.Older.AssumeNotNull());
+                break;
+
+            default:
+                Debug.Fail("A new ProjectChangeKind has been added that the RazorProjectInfoEndpointPublisher doesn't know how to deal with");
+                break;
+        }
+    }
+
+    // We want to wait until our project has been initialized (ProjectWorkspaceState != null)
+    // so that we don't publish half-finished projects, which can cause things like Semantic coloring to "flash"
+    // when they update repeatedly as they load.
+    private static bool ProjectWorkspacePublishable(IProjectSnapshot? project)
+    {
+        return project?.ProjectWorkspaceState != null;
```

Comment:
`ProjectWorkspaceState` is not nullable, so this is really just a null check on `project`. Consider removing the method entirely, and just inlining the null check as appropriate. eg, it shouldn't be necessary at all in the `StartSending` method

Reply:
Interesting, copied that helper from the file-based publisher and the comment seemed to imply it was possible to have null workspace state which indicated "not fully initialized project". I assume the design changed and it's no longer possible? Agreed that field is not nullable anymore, so I'll take your word for it. :) Removed that method. I assume file-based publisher will (hopefully) get deleted soon, so not cleaning up the code there.

Reply:
It was made not-nullable more recently, so that comment was out-of-date.